In [1]:
import  pandas as pd 
import numpy as np
import re


In [2]:
# Load the datasets
df_2014 = pd.read_csv("Data/OSMI Mental Health in Tech Survey 2014.csv", low_memory=False)
df_2016 = pd.read_csv("Data/OSMI mental-heath-in-tech-2016_20161114.csv", low_memory=False)
df_2017 = pd.read_csv("Data/OSMI Mental Health in Tech Survey 2017.csv", low_memory=False)
df_2018 = pd.read_csv("Data/OSMI Mental Health in Tech Survey 2018.csv", low_memory=False)
df_2019 = pd.read_csv("Data/OSMI 2019 Mental Health in Tech Survey Results.csv", low_memory=False)
df_2020 = pd.read_csv("Data/OSMI 2020 Mental Health in Tech Survey Results .csv", low_memory=False)
df_2021 = pd.read_csv("Data/OSMI 2021 Mental Health in Tech Survey Results .csv", low_memory=False)
df_2022 = pd.read_csv("Data/OSMI Mental Health in Tech Survey 2022.csv", low_memory=False)
df_2023 = pd.read_csv("Data/OSMI Mental Health in Tech Survey 2023.csv", low_memory=False)


In [3]:
# Display the first few rows of each dataset to understand their structure
df_2014.head()

,Timestamp,Age,Gender,Country,state,self_employed,family_history,treatment,work_interfere,no_employees,...,leave,mental_health_consequence,phys_health_consequence,coworkers,supervisor,mental_health_interview,phys_health_interview,mental_vs_physical,obs_consequence,comments
0,2014-08-27 11:29:31,37,Female,United States,IL,NaN,No,Yes,Often,6-25,...,Somewhat easy,No,No,Some of them,Yes,No,Maybe,Yes,No,NaN
1,2014-08-27 11:29:37,44,M,United States,IN,NaN,No,No,Rarely,More than 1000,...,Don't know,Maybe,No,No,No,No,No,Don't know,No,NaN
2,2014-08-27 11:29:44,32,Male,Canada,NaN,NaN,No,No,Rarely,6-25,...,Somewhat difficult,No,No,Yes,Yes,Yes,Yes,No,No,NaN
3,2014-08-27 11:29:46,31,Male,United Kingdom,NaN,NaN,Yes,Yes,Often,26-100,...,Somewhat difficult,Yes,Yes,Some of them,No,Maybe,Maybe,No,Yes,NaN
4,2014-08-27 11:30:22,31,Male,United States,TX,NaN,No,No,Never,100-500,...,Don't know,No,No,Some of them,Yes,Yes,Yes,Don't know,No,NaN


In [4]:
df_2016.head()

,Are you self-employed?,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health concerns and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:",...,"If you have a mental health issue, do you feel that it interferes with your work when being treated effectively?","If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?",What is your age?,What is your gender?,What country do you live in?,What US state or territory do you live in?,What country do you work in?,What US state or territory do you work in?,Which of the following best describes your work position?,Do you work remotely?
0,0,26-100,1.0,NaN,Not eligible for coverage / N/A,NaN,No,No,I don't know,Very easy,...,Not applicable to me,Not applicable to me,39,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Sometimes
1,0,6-25,1.0,NaN,No,Yes,Yes,Yes,Yes,Somewhat easy,...,Rarely,Sometimes,29,male,United States of America,Illinois,United States of America,Illinois,Back-end Developer|Front-end Developer,Never
2,0,6-25,1.0,NaN,No,NaN,No,No,I don't know,Neither easy nor difficult,...,Not applicable to me,Not applicable to me,38,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Always
3,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Sometimes,Sometimes,43,male,United Kingdom,NaN,United Kingdom,NaN,Supervisor/Team Lead,Sometimes
4,0,6-25,0.0,1.0,Yes,Yes,No,No,No,Neither easy nor difficult,...,Sometimes,Sometimes,43,Female,United States of America,Illinois,United States of America,Illinois,Executive Leadership|Supervisor/Team Lead|Dev ...,Sometimes


In [5]:
df_2017.head()

,#,<strong>Are you self-employed?</strong>,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,What is your gender?,What country do you <strong>live</strong> in?,What US state or territory do you <strong>live</strong> in?,What is your race?,Other.3,What country do you <strong>work</strong> in?,What US state or territory do you <strong>work</strong> in?,Start Date (UTC),Submit Date (UTC),Network ID
0,e49fe87572831232dcfa51b376b22039,0,100-500,1.0,1.0,No,Yes,No,I don't know,I don't know,...,Female,United Kingdom,NaN,NaN,NaN,United Kingdom,NaN,2018-05-16 12:32:04,2018-05-16 12:42:40,464b7a12f1
1,a1eede444ac024928d247a8372d54931,0,100-500,1.0,1.0,Yes,Yes,No,No,I don't know,...,male,United Kingdom,NaN,NaN,NaN,United Kingdom,NaN,2018-05-16 12:31:13,2018-05-16 12:40:40,464b7a12f1
2,37d3fd67f62bd1e0a2dea4f9cd440d98,0,6-25,1.0,1.0,I don't know,No,I don't know,No,Yes,...,male,United States of America,Missouri,White,NaN,United States of America,Missouri,2018-05-09 05:34:05,2018-05-09 05:46:04,1eb7e0cb94
3,519b759442c1cab0e9b5a8a1acb1b216,0,More than 1000,1.0,1.0,Yes,Yes,I don't know,I don't know,Yes,...,Male,United States of America,Washington,White,NaN,United States of America,Washington,2018-05-04 23:19:14,2018-05-04 23:23:23,63852edbc4
4,ef0af4927b575b1a3e607c11ca37870e,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,female,United States of America,Illinois,More than one of the above,NaN,United States of America,Illinois,2018-05-03 00:40:24,2018-05-03 00:53:20,43237889f1


In [6]:
df_2018.head()

,#,<strong>Are you self-employed?</strong>,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,What is your gender?,What country do you <strong>live</strong> in?,What US state or territory do you <strong>live</strong> in?,What is your race?,Other.3,What country do you <strong>work</strong> in?,What US state or territory do you <strong>work</strong> in?,Start Date (UTC),Submit Date (UTC),Network ID
0,e44a0a34f2465940beda2a1537e9b99e,0,More than 1000,1.0,0.0,Yes,Yes,Yes,Yes,Yes,...,Female,Canada,NaN,NaN,NaN,Canada,NaN,2018-12-29 23:46:38,2018-12-30 00:00:03,4bbb884ccc
1,0d698e3beca20fb75f19b9d528e36d73,0,More than 1000,1.0,1.0,Yes,Yes,No,I don't know,I don't know,...,male,United States of America,Massachusetts,White,NaN,United States of America,Massachusetts,2018-12-27 21:40:40,2018-12-27 21:45:45,275e7543bd
2,61a40c9071eb36fa9caa254d31500c41,0,6-25,0.0,1.0,Yes,Yes,No,No,I don't know,...,Male,United States of America,Florida,White,NaN,United States of America,Florida,2018-12-21 17:37:43,2018-12-21 18:08:01,43994c3dba
3,f8624340bead7deb08abb766704ddf6b,0,6-25,1.0,1.0,No,No,No,No,I don't know,...,male,Norway,NaN,NaN,NaN,Norway,NaN,2018-12-21 16:37:56,2018-12-21 16:44:44,907b3a3faa
4,31d3ae93b68d79e504a0a643601b6b1e,0,26-100,1.0,1.0,Yes,Yes,Yes,Yes,Yes,...,Ostensibly Male,United States of America,Tennessee,White,NaN,United States of America,Tennessee,2018-12-20 19:39:02,2018-12-20 20:58:34,26df20fea3


In [7]:
df_2019.head()

,*Are you self-employed?*,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, how easy or difficult would it be to ask for that leave?",...,Briefly describe what you think the industry as a whole and/or employers could do to improve mental health support for employees.,"If there is anything else you would like to tell us that has not been covered by the survey questions, please use this space to do so.",Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used _anonymously_ and only with your permission.),What is your age?,What is your gender?,What country do you *live* in?,What US state or territory do you *live* in?,What is your race?,What country do you *work* in?,What US state or territory do you *work* in?
0,False,26-100,True,True,I don't know,No,Yes,Yes,I don't know,Very easy,...,NaN,NaN,False,25,Male,United States of America,Nebraska,White,United States of America,Nebraska
1,False,26-100,True,True,Yes,No,No,Yes,Yes,I don't know,...,NaN,NaN,False,51,male,United States of America,Nebraska,White,United States of America,Nebraska
2,False,26-100,True,True,I don't know,No,No,I don't know,I don't know,Somewhat difficult,...,I think opening up more conversation around th...,Thank you,True,27,Male,United States of America,Illinois,White,United States of America,Illinois
3,False,100-500,True,True,I don't know,No,Yes,Yes,Yes,Very easy,...,NaN,NaN,False,37,male,United States of America,Nebraska,White,United States of America,Nebraska
4,False,26-100,True,True,I don't know,No,I don't know,I don't know,I don't know,I don't know,...,NaN,NaN,False,46,m,United States of America,Nebraska,White,United States of America,Nebraska


In [8]:
df_2020.head()

,#,*Are you self-employed?*,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,"If there is anything else you would like to tell us that has not been covered by the survey questions, please use this space to do so.",Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used _anonymously_ and only with your permission.),What is your age?,What is your gender?,What country do you *live* in?,What US state or territory do you *live* in?,What is your race?,Other.3,What country do you *work* in?,What US state or territory do you *work* in?
0,zwrffw6ykfo82ft1twvzwrffw6c6wsfv,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,0,45,Male,United States of America,Connecticut,White,NaN,United States of America,Connecticut
1,zhdmhaa8r0125c4zmoi7qzhdmtjrakhm,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,1,24,female,Russia,NaN,NaN,NaN,Russia,NaN
2,x4itwa9hnlw7qke4y5xibx4itwa9yzl5,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,mental health should be a law by government.,1,46,Male,India,NaN,NaN,NaN,India,NaN
3,x3v3oimu5pn0043n8x3v3oizaybhwwto,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,1,25,Female,Canada,NaN,NaN,NaN,Canada,NaN
4,uyp6re7bhnyx6gez09uyp6re72z0e4e4,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,no,1,25,F,Canada,NaN,NaN,NaN,Canada,NaN


In [9]:
df_2021.head()

,#,*Are you self-employed?*,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,What country do you *live* in?,What US state or territory do you *live* in?,What is your race?,Other.3,What country do you *work* in?,What US state or territory do you *work* in?,Have you been diagnosed with COVID-19?,Start Date (UTC),Submit Date (UTC),Network ID
0,eu6sxt9r6pyxxykpxcqs6r2qeu6sxt9r,0,26-100,1.0,1.0,I don't know,No,No,I don't know,I don't know,...,United States of America,Alaska,White,NaN,United States of America,Alaska,NaN,2021-12-16 20:28:29,2021-12-16 20:36:02,f3adc1b78a
1,mqgsjpmc31k0fuymqgv96vapalozx3q0,0,500-1000,1.0,1.0,Yes,No,Yes,Yes,Yes,...,Brazil,NaN,NaN,NaN,Brazil,NaN,NaN,2021-12-16 13:42:51,2021-12-16 18:44:40,49a51b7ac1
2,fkuijv5hpifz4fusbn1fkuib4ipqt19s,0,100-500,1.0,1.0,Yes,Yes,No,I don't know,I don't know,...,Brazil,NaN,NaN,NaN,Brazil,NaN,NaN,2021-12-16 13:01:19,2021-12-16 13:14:49,f9e26fd31a
3,eb7i5ye61yizgs5veb7i5yyttqp5zwp2,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Italy,NaN,NaN,NaN,Italy,NaN,NaN,2021-12-16 08:57:11,2021-12-16 08:59:00,73a0e1b4c5
4,0qoaw3bb3kzbb2oet0qoa3xc3yzdrvva,0,More than 1000,0.0,1.0,Yes,No,Yes,Yes,I don't know,...,Canada,NaN,NaN,NaN,Canada,NaN,NaN,2021-12-08 14:35:27,2021-12-08 14:44:10,bade27e84d


In [10]:
df_2022.head()

,#,*Are you self-employed?*,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,What is your race?,Other.3,What country do you *work* in?,What US state or territory do you *work* in?,Have you been diagnosed with COVID-19?,Response Type,Start Date (UTC),Submit Date (UTC),Network ID,Tags
0,454yabpvj5ekod27qm454yabpu6ww4fa,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,Russia,NaN,NaN,completed,2022-12-25 13:04:39,2022-12-25 13:12:15,16f9edd3e7,NaN
1,turfs5uwaygy96omjyvmxturfaq4tqpm,0,26-100,1.0,1.0,I don't know,No,No,No,I don't know,...,NaN,NaN,Netherlands,NaN,NaN,completed,2022-12-21 17:47:17,2022-12-21 18:09:51,af98dee606,NaN
2,1hb89b2fyz03jg1hdvv3unu46hmug4lb,0,26-100,1.0,1.0,No,Yes,I don't know,I don't know,Yes,...,NaN,NaN,Sweden,NaN,NaN,completed,2022-12-14 12:40:16,2022-12-14 12:48:02,e223d877cb,NaN
3,88shtlfnzinq2nrbw1ib88shju9q7xxh,0,More than 1000,1.0,1.0,Yes,Yes,Yes,Yes,I don't know,...,White,NaN,United States of America,Oregon,NaN,completed,2022-12-07 15:46:30,2022-12-07 16:25:24,b2e27a741e,NaN
4,8y9n53igp152mvon6woisdiq8y9n53ig,0,26-100,1.0,1.0,Yes,No,Yes,Yes,Yes,...,NaN,NaN,United Kingdom,NaN,NaN,completed,2022-12-01 12:23:51,2022-12-01 12:33:14,1d324732b0,NaN


In [11]:
df_2023.head()

,#,*Are you self-employed?*,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided health coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health disorders and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,...,What is your race?,Other.3,What country do you *work* in?,What US state or territory do you *work* in?,Have you been diagnosed with COVID-19?,Response Type,Start Date (UTC),Submit Date (UTC),Network ID,Tags
0,an587p4oi8h7gtq9an5874mxwixvx5p6,0,26-100,1.0,0.0,No,No,No,No,Yes,...,More than one of the above,NaN,United States of America,Colorado,NaN,completed,2023-03-03 22:04:39,2023-03-05 23:50:09,6dde6a8348,NaN
1,2pivyqqnhwt0b7lp2pivythv3eonu6me,0,6-25,1.0,1.0,I don't know,No,No,I don't know,I don't know,...,White,NaN,United States of America,Indiana,NaN,completed,2023-03-02 22:20:33,2023-03-02 22:31:00,90e5ba7a6a,NaN
2,06ra2i3glzxtm8m2s06ra2z5vmep483a,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,White,NaN,United States of America,Maryland,NaN,completed,2023-03-02 20:45:20,2023-03-02 21:06:03,1a0014e801,NaN
3,4ojrsw9n2x8gr9wabb24ojrswr9znb1o,0,500-1000,0.0,0.0,Yes,Yes,Yes,No,I don't know,...,White,NaN,United States of America,Indiana,NaN,completed,2023-03-02 17:52:57,2023-03-02 18:06:01,92ea1836e3,NaN
4,wuks4mzrqdbuqkwm8hwuks5yvhjd6p58,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,White,NaN,United States of America,Ohio,NaN,completed,2023-03-02 13:53:59,2023-03-02 14:11:32,94b54112a3,NaN


In [12]:
print(df_2014.shape)
print(df_2016.shape)
print(df_2017.shape)
print(df_2018.shape)
print(df_2019.shape)
print(df_2020.shape)
print(df_2021.shape)
print(df_2022.shape)
print(df_2023.shape)

(1259, 27)
(1433, 63)
(756, 123)
(417, 123)
(352, 82)
(180, 120)
(131, 124)
(164, 126)
(6, 126)


In [13]:
#Display the column names of each dataset to identify commonalities and differences

print("2014 Columns:")
print(df_2014.columns.tolist(), "\n")

print("2016 Columns:")
print(df_2016.columns.tolist(), "\n")

print("2017 Columns:")
print(df_2017.columns.tolist(), "\n")

print("2018 Columns:")
print(df_2018.columns.tolist(), "\n")

print("2019 Columns:")
print(df_2019.columns.tolist(), "\n")

print("2020 Columns:")
print(df_2020.columns.tolist(), "\n")

print("2021 Columns:")
print(df_2021.columns.tolist(), "\n")

print("2022 Columns:")
print(df_2022.columns.tolist(), "\n")

print("2023 Columns:")
print(df_2023.columns.tolist(), "\n") 


2014 Columns:
['Timestamp', 'Age', 'Gender', 'Country', 'state', 'self_employed', 'family_history', 'treatment', 'work_interfere', 'no_employees', 'remote_work', 'tech_company', 'benefits', 'care_options', 'wellness_program', 'seek_help', 'anonymity', 'leave', 'mental_health_consequence', 'phys_health_consequence', 'coworkers', 'supervisor', 'mental_health_interview', 'phys_health_interview', 'mental_vs_physical', 'obs_consequence', 'comments'] 

2016 Columns:
['Are you self-employed?', 'How many employees does your company or organization have?', 'Is your employer primarily a tech company/organization?', 'Is your primary role within your company related to tech/IT?', 'Does your employer provide mental health benefits as part of healthcare coverage?', 'Do you know the options for mental health care available under your employer-provided coverage?', 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?', '

### Data Cleaning and  Preprocessing 

In [14]:
# Standardized column mapping for 2014–2023 OSMI datasets
column_mapping = {
    "age": ["Age", "What is your age?"],
    "gender": ["Gender", "What is your gender?"],
    "country": ["Country", "What country do you *live* in?"],
    "company_size": [
        "no_employees",
        "How many employees does your company or organization have?"
    ],
    "family_history": [
        "family_history",
        "Do you have a family history of mental illness?"
    ],
    "benefits": [
        "benefits",
        "Does your employer provide mental health benefits as part of healthcare coverage?"
    ],
    "care_options": [
        "care_options",
        "Do you know the options for mental health care available under your employer-provided health coverage?"
    ],
    "work_interfere": [
        "work_interfere",
        "If you have a mental health disorder, how often do you feel that it interferes with your work *when* _*NOT*_* being treated effectively (i.e., when you are experiencing symptoms)?*"
    ],
    "anonymity": [
        "anonymity",
        "Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?"
    ],
    "leave": [
        "leave",
        "If a mental health issue prompted you to request a medical leave from work, how easy or difficult would it be to ask for that leave?"
    ],
    "treatment": [
        "treatment",
        "Have you ever sought treatment for a mental health disorder from a mental health professional?"
    ]
}
  



In [15]:
# Function to rename columns based on the mapping
def clean_osmi(df, year, mapping):
    # Start a new clean dataframe
    new_df = pd.DataFrame()
    new_df["year"] = year

    for clean_name, possible_names in mapping.items():
        found = False
        
        for col in possible_names:
            if col in df.columns:
                new_df[clean_name] = df[col]
                found = True
                break
        
        # if no matching column was found in this year's dataset
        if not found:
            new_df[clean_name] = None
    
    return new_df




In [16]:
# Clean each year's dataset
df_2014_clean = clean_osmi(df_2014, 2014, column_mapping)
df_2016_clean = clean_osmi(df_2016, 2016, column_mapping)
df_2017_clean = clean_osmi(df_2017, 2017, column_mapping)
df_2018_clean = clean_osmi(df_2018, 2018, column_mapping)
df_2019_clean = clean_osmi(df_2019, 2019, column_mapping)
df_2020_clean = clean_osmi(df_2020, 2020, column_mapping)
df_2021_clean = clean_osmi(df_2021, 2021, column_mapping)
df_2022_clean = clean_osmi(df_2022, 2022, column_mapping)
df_2023_clean = clean_osmi(df_2023, 2023, column_mapping)

In [17]:
# Combine all cleaned dataframes into a single dataframe
combined_df = pd.concat([
    df_2014_clean, df_2016_clean, df_2017_clean, df_2018_clean,
    df_2019_clean, df_2020_clean, df_2021_clean, df_2022_clean, df_2023_clean
], ignore_index=True)



In [18]:
# Explore the combined dataframe
combined_df.head()



,year,age,gender,country,company_size,family_history,benefits,care_options,work_interfere,anonymity,leave,treatment
0,NaN,37.0,Female,United States,6-25,No,Yes,Not sure,Often,Yes,Somewhat easy,Yes
1,NaN,44.0,M,United States,More than 1000,No,Don't know,No,Rarely,Don't know,Don't know,No
2,NaN,32.0,Male,Canada,6-25,No,No,No,Rarely,Don't know,Somewhat difficult,No
3,NaN,31.0,Male,United Kingdom,26-100,Yes,No,Yes,Often,No,Somewhat difficult,Yes
4,NaN,31.0,Male,United States,100-500,No,Yes,No,Never,Don't know,Don't know,No


In [19]:
combined_df.shape


(4698, 12)

In [20]:
combined_df.isna().sum()

year              4698
age                  2
gender              34
country           2606
company_size       588
family_history       0
benefits          1231
care_options      1908
work_interfere    2870
anonymity          588
leave             1734
treatment         1433
dtype: int64

#### **Year**

- While combining surveys from multiple years, the year column didn’t attach cleanly. To fix it, unique values were checked and each dataset was assigned its correct year before merging.

In [21]:
combined_df["year"].unique()


array([nan])

In [22]:
# Manually assign years based on known row indices
combined_df.loc[0:1258, "year"] = 2014
combined_df.loc[1259:2691, "year"] = 2016
combined_df.loc[2692:3447, "year"] = 2017
combined_df.loc[3448:3864, "year"] = 2018
combined_df.loc[3865:4216, "year"] = 2019
combined_df.loc[4217:4396, "year"] = 2020
combined_df.loc[4397:4527, "year"] = 2021
combined_df.loc[4528:4691, "year"] = 2022
combined_df.loc[4692:4697, "year"] = 2023


In [23]:
combined_df["year"].unique()

array([2014., 2016., 2017., 2018., 2019., 2020., 2021., 2022., 2023.])

In [24]:
# Explore the combined dataframe to confirm changes
combined_df.head()

,year,age,gender,country,company_size,family_history,benefits,care_options,work_interfere,anonymity,leave,treatment
0,2014.0,37.0,Female,United States,6-25,No,Yes,Not sure,Often,Yes,Somewhat easy,Yes
1,2014.0,44.0,M,United States,More than 1000,No,Don't know,No,Rarely,Don't know,Don't know,No
2,2014.0,32.0,Male,Canada,6-25,No,No,No,Rarely,Don't know,Somewhat difficult,No
3,2014.0,31.0,Male,United Kingdom,26-100,Yes,No,Yes,Often,No,Somewhat difficult,Yes
4,2014.0,31.0,Male,United States,100-500,No,Yes,No,Never,Don't know,Don't know,No


#### **Age**

- The age column contained unrealistic values and missing entries. Ages below 15 or above 80 were treated as errors and replaced with the median age, keeping the data realistic

In [25]:
# Convert age to numbers, invalid entries -> NaN
combined_df["age"] = pd.to_numeric(combined_df["age"], errors="coerce")

In [26]:
# Set unrealistic ages to NaN
combined_df.loc[(combined_df["age"] < 18) | (combined_df["age"] > 100), "age"] = np.nan

In [27]:
# Replace missing ages using median
combined_df["age"].fillna(combined_df["age"].median(), inplace=True)

In [28]:
# Confirm age cleaned correctly
combined_df["age"].describe()

count    4698.000000
mean       34.041933
std         8.379476
min        18.000000
25%        28.000000
50%        33.000000
75%        39.000000
max        99.000000
Name: age, dtype: float64

In [29]:
combined_df["age"].isna().sum()

0

#### **Gender**

- The *Gender* column had dozens of messy text variations. To make the model readable, all responses were standardized into three groups: Male, Female, and Other.

In [30]:
# Clean Gender Column

# Convert gender values to lowercase and strip spaces
combined_df['gender'] = combined_df['gender'].astype(str).str.lower().str.strip()

# Replace common male variations with "Male"
male_terms = ['male', 'm', 'man', 'cis male', 'male-ish', 'msle', 'mal', 'maile']
combined_df['gender'] = combined_df['gender'].replace(male_terms, 'Male')

# Replace common female variations with "Female"
female_terms = ['female', 'f', 'woman', 'cis female', 'femail']
combined_df['gender'] = combined_df['gender'].replace(female_terms, 'Female')

# Everything else → Other
combined_df.loc[
    ~combined_df['gender'].isin(['Male', 'Female']),
    'gender'
] = 'Other'

# Check distribution after cleaning
combined_df['gender'].value_counts()


gender
Male      3313
Female    1150
Other      235
Name: count, dtype: int64

#### **Country**

- Countries appeared in many different text formats and inconsistent entries. To simplify modeling, country names were standardized, and rare countries were grouped together into an “Other” category to reduce noise.

In [31]:
# --- Clean Country Column ---

# Convert to string, strip spaces, and title-case
combined_df['country'] = combined_df['country'].astype(str).str.strip().str.title()

# Replace common variations
combined_df['country'] = combined_df['country'].replace({
    'Usa': 'United States',
    'United States Of America': 'United States',
    'Uk': 'United Kingdom',
    'U.S.A.': 'United States'
})

# Replace meaningless or unknown entries
combined_df.loc[
    combined_df['country'].isin(['Na', 'None', 'Nan', '', 'Not Applicable']),
    'country'
] = np.nan

# Check unique values after cleaning
combined_df['country'].nunique()


81

- Countries appeared in 81 variations and low-frequency categories. To keep the model stable, the top 10 countries were kept, and the rest were grouped as ‘Other’

In [32]:
# Group countries into top N and label rest 'Other'

# Count frequencies
top_n = 10
country_counts = combined_df['country'].value_counts(dropna=True)

# Create mapping: top N keep as-is, others -> 'Other'
top_countries = country_counts.nlargest(top_n).index.tolist()
combined_df['country_grouped'] = combined_df['country'].where(combined_df['country'].isin(top_countries), other='Other')

# Show results
print("Top countries kept:", top_countries)
print("Unique in country_grouped:", combined_df['country_grouped'].nunique())
print(combined_df['country_grouped'].value_counts().head(15))


Top countries kept: ['United States', 'United Kingdom', 'Canada', 'India', 'Germany', 'Netherlands', 'Brazil', 'Australia', 'Ireland', 'Portugal']
Unique in country_grouped: 11
country_grouped
Other             2887
United States     1138
United Kingdom     241
Canada             113
India               84
Germany             68
Netherlands         43
Brazil              39
Australia           32
Ireland             31
Portugal            22
Name: count, dtype: int64


#### **Company Size**

- Company size had inconsistent formats like ‘6 - 25’, ‘more than 1000’, and even text responses. These were normalized into clean size ranges and converted into ordered categories.

In [33]:
# Normalize company size text and convert to ordered numeric categories

# First, normalize text
combined_df['company_size_raw'] = combined_df['company_size'].astype(str).str.strip()

# Map common variants to canonical labels
size_map = {
    '1-5': '1-5', '6-25': '6-25', '26-100': '26-100', '100-500': '100-500',
    '500-1000': '500-1000', 'more than 1000': '1000+', 'more than 1000 ': '1000+',
    'more than 1000': '1000+', '1000+': '1000+', '6 - 25': '6-25'
}

combined_df['company_size_clean'] = combined_df['company_size_raw'].replace(size_map)

# Define function to infer size from messy text 
def infer_size(x):
    if pd.isna(x) or x.lower() in ['nan', 'none', 'nan']:
        return np.nan
    s = str(x)
    # simple rules
    if re.search(r'1000|more than 1000|>1000', s, flags=re.I):
        return '1000+'
    if re.search(r'500', s):
        return '500-1000'
    if re.search(r'100-500', s):
        return '100-500'
    if re.search(r'26-100', s):
        return '26-100'
    if re.search(r'6-25|6 - 25', s):
        return '6-25'
    if re.search(r'1-5|1 - 5|1 to 5', s):
        return '1-5'
    return np.nan

combined_df['company_size_clean'] = combined_df['company_size_clean'].fillna(combined_df['company_size_raw'].apply(infer_size))

# Create an ordered numeric version for modeling (small -> large)
order = {'1-5':1, '6-25':2, '26-100':3, '100-500':4, '500-1000':5, '1000+':6}
combined_df['company_size_ord'] = combined_df['company_size_clean'].map(order)

# Check results
print(combined_df['company_size_clean'].value_counts(dropna=False))
print("Missing company_size:", combined_df['company_size_ord'].isna().sum())



company_size_clean
More than 1000    1064
26-100             909
100-500            859
6-25               729
nan                588
500-1000           287
1-5                262
Name: count, dtype: int64
Missing company_size: 1652


In [34]:
# Add missing mapping to fix missing company size 
combined_df['company_size_clean'] = combined_df['company_size_clean'].replace({
    'More than 1000': '1000+',
    'more than 1000': '1000+'
})

# Re-map
combined_df['company_size_ord'] = combined_df['company_size_clean'].map(order)

# Check results 
print(combined_df['company_size_clean'].value_counts(dropna=False))
print("Missing company_size:", combined_df['company_size_ord'].isna().sum())


company_size_clean
1000+       1064
26-100       909
100-500      859
6-25         729
nan          588
500-1000     287
1-5          262
Name: count, dtype: int64
Missing company_size: 588


- Converted to an ordered scale and filled missing values with the median category.

In [35]:
# Fill missing company size ordinals with median to clean company_size_ord
combined_df['company_size_ord'].fillna(combined_df['company_size_ord'].median(), inplace=True)
combined_df['company_size_ord'].isna().sum()


0

#### **Family History**

- This column contained different text formats for yes/no responses. All variations were mapped into a simple numeric binary (1 = Yes, 0 = No) so the model can use it.

In [ ]:
# --- Family history to binary (Yes -> 1, No -> 0) ---

# Convert family_history to numeric binary 1/0
combined_df['family_history_clean'] = combined_df['family_history'].astype(str).str.lower().str.strip()
combined_df['family_history_clean'] = combined_df['family_history_clean'].replace({
    'yes':'1', 'no':'0', 'y':'1', 'n':'0', 'true':'1', 'false':'0'
})
combined_df['family_history'] = combined_df['family_history_clean'].map({'1':1, '0':0}).astype('float')
print("family_history value counts:", combined_df['family_history'].value_counts(dropna=False))



family_history value counts: family_history
1.0    2026
0.0    1936
NaN     736
Name: count, dtype: int64


- To deal with the 736 NaNs, fill missing family history with the mode 

In [ ]:
# Impute missing family_history with mode
combined_df['family_history'].fillna(combined_df['family_history'].mode()[0], inplace=True)

# Confirm the imputation 
combined_df['family_history'].value_counts()


family_history
1.0    2762
0.0    1936
Name: count, dtype: int64

#### **Benefits. Care Options . Anonymity**

- These three columns recorded Yes/No-style responses, but with many messy variations (e.g., “Y”, “Nope”, “Don’t know”). They were cleaned using the same process:

- Standardized all variations of yes and no into numeric 1/0.

- Marked uncertain answers (like “don’t know”) as missing.

- Filled missing values using the most common response (mode).

In [ ]:
# --- Standardize yes/no/dont know type columns ---

# Map Yes->1, No->0, uncertain answers -> NaN for key binary columns and target

binary_cols = ['benefits', 'care_options', 'anonymity']  

def map_yesno(x):
    if pd.isna(x):
        return np.nan
    s = str(x).strip().lower()
    if s in ['yes', 'y', 'true']:
        return 1
    if s in ['no', 'n', 'false']:
        return 0
    if s in ["don't know", "dont know", "not sure", "i don't know", "do not know"]:
        return np.nan  # treat as unknown
    return np.nan

for c in binary_cols:
    combined_df[c + '_bin'] = combined_df[c].apply(map_yesno)


print("benefits_bin counts:\n", combined_df['benefits_bin'].value_counts(dropna=False))
print("care_options_bin counts:\n", combined_df['care_options_bin'].value_counts(dropna=False))
print("anonymity_bin counts:\n", combined_df['anonymity_bin'].value_counts(dropna=False))


benefits_bin counts:
 benefits_bin
NaN    2368
1.0    1571
0.0     759
Name: count, dtype: int64
care_options_bin counts:
 care_options_bin
NaN    2222
0.0    1317
1.0    1159
Name: count, dtype: int64
anonymity_bin counts:
 anonymity_bin
NaN    3163
1.0    1288
0.0     247
Name: count, dtype: int64


In [42]:
# --- Impute missing binary columns with mode ---
binary_clean_cols = ['benefits_bin', 'care_options_bin', 'anonymity_bin']

for col in binary_clean_cols:
    combined_df[col].fillna(combined_df[col].mode()[0], inplace=True)

# Check results
for col in binary_clean_cols:
    print(f"{col} after imputation:\n", combined_df[col].value_counts(dropna=False), "\n")


benefits_bin after imputation:
 benefits_bin
1.0    3939
0.0     759
Name: count, dtype: int64 

care_options_bin after imputation:
 care_options_bin
0.0    3539
1.0    1159
Name: count, dtype: int64 

anonymity_bin after imputation:
 anonymity_bin
1.0    4451
0.0     247
Name: count, dtype: int64 



### **Treatment**

- Only “Yes” and “No” were kept and converted into 1/0. Where 1 = sought treatment and 0 = did not.

In [43]:
# Treatment (target) mapping: Yes -> 1, No -> 0, others -> NaN
combined_df['treatment_clean'] = combined_df['treatment'].astype(str).str.strip().str.lower().map({
    'yes':1, 'y':1, 'true':1, 'no':0, 'n':0, 'false':0
})

print("treatment_clean counts:\n", combined_df['treatment_clean'].value_counts(dropna=False))

treatment_clean counts:
 treatment_clean
NaN    3087
1.0     854
0.0     757
Name: count, dtype: int64


- Rows with missing treatment were removed to ensure correct labels.

In [ ]:
# Drop rows with missing treatment (target) 
print("Missing treatment before drop:", combined_df['treatment_clean'].isna().sum())

combined_df = combined_df.dropna(subset=['treatment_clean'])

print("New shape after dropping missing target:", combined_df.shape)
print("Treatment distribution after cleaning:\n", combined_df['treatment_clean'].value_counts())


Missing treatment before drop: 3087
New shape after dropping missing target: (1611, 21)
Treatment distribution after cleaning:
 treatment_clean
1.0    854
0.0    757
Name: count, dtype: int64


#### **Work Interfere**

- Converted “Never/Rarely/Sometimes/Often” into an ordered numeric scale then the missing values were filled using the median, preserving the ordinal nature

In [ ]:
# Clean work_interfere into ordinal

# Map work_interfere responses to ordinal values
map_work = {
    'never': 0,
    'rarely': 1,
    'sometimes': 2,
    'often': 3
}

combined_df['work_interfere_clean'] = combined_df['work_interfere'].astype(str).str.lower().str.strip().map(map_work)

# Impute missing with median
combined_df['work_interfere_ord'] = combined_df['work_interfere_clean'].fillna(combined_df['work_interfere_clean'].median())

print("work_interfere_ord:\n", combined_df['work_interfere_ord'].value_counts())


work_interfere_ord:
 work_interfere_ord
2.0    915
3.0    293
0.0    217
1.0    186
Name: count, dtype: int64


#### **Leave** 

- Mapped leave-difficulty categories into an ordered scale from easy → difficult. To keep this order, categories were converted into ordered integers. Missing values were filled with the most common category(mode).

In [ ]:
#Clean leave into ordinal

# Map leave responses to ordinal values
map_leave = {
    'very easy': 0,
    'somewhat easy': 1,
    "don't know": 2,
    'somewhat difficult': 3,
    'very difficult': 4
}

combined_df['leave_clean'] = combined_df['leave'].astype(str).str.lower().str.strip().map(map_leave)

# Impute missing leave values with mode
combined_df['leave_ord'] = combined_df['leave_clean'].fillna(combined_df['leave_clean'].mode()[0])

print("leave_ord:\n", combined_df['leave_ord'].value_counts())


leave_ord:
 leave_ord
2.0    713
1.0    349
0.0    276
3.0    175
4.0     98
Name: count, dtype: int64
